src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,431 @@
+            for (uint i = 0; i < newBlocksNeeded; i++)
+            {
+                Debug.Assert(_blocks[allocatedBlockCount] is null);
+                _blocks[allocatedBlockCount++] = ArrayPool<byte>.Shared.Rent(BlockSize);
+            }
+
+            _allocatedEnd = allocatedBlockCount * BlockSize;
+
+            // After all of that, we should have enough available memory now
+            Debug.Assert(byteCount <= AvailableMemory.Length);
+
+            CheckState();
+        }
+
+        [Conditional("DEBUG")]
+        private void CheckState()
+        {
+            if (_blocks == null)
+            {
+                Debug.Assert(_activeStart == 0);
+                Debug.Assert(_availableStart == 0);
+                Debug.Assert(_allocatedEnd == 0);
+            }
+            else
+            {
+                Debug.Assert(_activeStart <= _availableStart);
+                Debug.Assert(_availableStart <= _allocatedEnd);
+                Debug.Assert(_allocatedEnd <= _blocks.Length * BlockSize);
+
+                Debug.Assert(_allocatedEnd % BlockSize == 0, $"_allocatedEnd={_allocatedEnd} not at block boundary?");
+
+                uint firstAllocatedBlock = _activeStart / BlockSize;
+                uint firstUnallocatedBlock = _allocatedEnd / BlockSize;
+
+                for (uint i = 0; i < firstAllocatedBlock; i++)
+                {
+                    Debug.Assert(_blocks[i] is null);
+                }
+
+                for (uint i = firstAllocatedBlock; i < firstUnallocatedBlock; i++)
+                {
+                    Debug.Assert(_blocks[i] is not null);
+                }
+
+                for (uint i = firstUnallocatedBlock; i < _blocks.Length; i++)
+                {
+                    Debug.Assert(_blocks[i] is null);
+                }
+
+                if (_activeStart == _availableStart)
+                {
+                    Debug.Assert(_activeStart == 0, $"No active bytes but _activeStart={_activeStart}");
+                }
+            }
+        }
+    }
+
+    // This is a Memory-like struct for handling multi-array segments from MultiArrayBuffer above.
+    // It supports standard Span/Memory operations like indexing, Slice, Length, etc
+    // It also supports CopyTo/CopyFrom Span<byte>
+
+    internal readonly struct MultiMemory
+    {
+        private readonly byte[]?[]? _blocks;
+        private readonly uint _start;
+        private readonly uint _length;
+
+        private const int BlockSize = 16 * 1024;
+
+        internal MultiMemory(byte[]?[]? blocks, uint start, uint length)
+        {
+            if (length == 0)
+            {
+                _blocks = null;
+                _start = 0;
+                _length = 0;
+            }
+            else
+            {
+                Debug.Assert(blocks is not null);
+                Debug.Assert(start >= 0);
+                Debug.Assert(length >= 0);
```

Comment:
With `uint` these will always be true.
Should check instead
```suggestion
                Debug.Assert(start <= int.MaxValue);
                Debug.Assert(length <= int.MaxValue);
```
?

(Should the compiler have shown a suggestion for this?)